### PR TITLE
Improve navigation mesh code examples

### PR DESCRIPTION
**Pull request by smix8**
Improves navigation mesh code examples.

Users had trouble adapting the example code snippets to their projects, struggling with how to make things reusable or with other  specifics.

The improved examples are now longer but show nearly the entire flow from start to finish how to create and update navigation meshes in scripts and mention some common problems with e.g. TileMap shenanigans or SceneTree processing.

### CODE REVIEW

File: tutorials/navigation/navigation_using_navigationmeshes.rst [hunk 1]
```diff
@@ -229,95 +229,152 @@ There are some common user problems and important caveats to consider when creat
     Remove the geometry that is on the ground inside the other geometry. If that is not possible, add smaller "dummy" geometry inside with as few triangles as possible so the cells
     are occupied with something.
 
+Navigation mesh script templates
+--------------------------------
+
+The following script uses the NavigationServer to parse source geometry from the SceneTree, bakes a navigation mesh, and updates a navigation region with the updated navigation mesh.
```
Comment (timothyqiu) [inline tutorials/navigation/navigation_using_navigationmeshes.rst:235]:
```suggestion
The following script uses the NavigationServer to parse source geometry from the scene tree, bakes a navigation mesh, and updates a navigation region with the updated navigation mesh.
```

It's not related to the `SceneTree` class, but about the scene tree concept.

File: tutorials/navigation/navigation_using_navigationmeshes.rst [hunk 2]
```diff
@@ -229,95 +229,152 @@ There are some common user problems and important caveats to consider when creat
...
+    var callback_parsing : Callable
+    var callback_baking : Callable
+    var region_rid: RID
+
+    func _ready() -> void:
+        navigation_mesh = NavigationPolygon.new()
+        navigation_mesh.agent_radius = 10.0
+        source_geometry = NavigationMeshSourceGeometryData2D.new()
+        callback_parsing = Callable(self, "on_parsing_done")
+        callback_baking = Callable(self, "on_baking_done")
```
Comment (timothyqiu) [inline tutorials/navigation/navigation_using_navigationmeshes.rst:253]:
```suggestion
        callback_parsing = on_parsing_done
        callback_baking = on_baking_done
```

Method is itself a Callable.

File: tutorials/navigation/navigation_using_navigationmeshes.rst [hunk 3]
```diff
@@ -229,95 +229,152 @@ There are some common user problems and important caveats to consider when creat
...
+        region_rid = NavigationServer2D.region_create()
+
+        # Enable the region and set it to the default navigation map.
+        NavigationServer2D.region_set_enabled(region_rid, true)
+        NavigationServer2D.region_set_map(region_rid, get_world_2d().get_navigation_map())
+
+        # Some mega-nodes like TileMap are often not ready on the first frame.
+        # Also the parsing needs to happen on the main-thread.
+        # So do a deferred call to avoid common parsing issues.
+        call_deferred("parse_source_geometry")
```
Comment (timothyqiu) [inline tutorials/navigation/navigation_using_navigationmeshes.rst:263]:
```suggestion
        parse_source_geometry.call_deferred()
```

Same here. Better avoid using strings :)

File: tutorials/navigation/navigation_using_navigationmeshes.rst [hunk 4]
```diff
@@ -229,95 +229,152 @@ There are some common user problems and important caveats to consider when creat
...
+    var callback_parsing : Callable
+    var callback_baking : Callable
+    var region_rid: RID
+
+    func _ready() -> void:
+        navigation_mesh = NavigationMesh.new()
+        navigation_mesh.agent_radius = 0.5
+        source_geometry = NavigationMeshSourceGeometryData3D.new()
+        callback_parsing = Callable(self, "on_parsing_done")
+        callback_baking = Callable(self, "on_baking_done")
```
Comment (timothyqiu) [inline tutorials/navigation/navigation_using_navigationmeshes.rst:314]:
```suggestion
        callback_parsing = on_parsing_done
        callback_baking = on_baking_done
```

Same as the 2D version.

File: tutorials/navigation/navigation_using_navigationmeshes.rst [hunk 5]
```diff
@@ -229,95 +229,152 @@ There are some common user problems and important caveats to consider when creat
...
+        region_rid = NavigationServer3D.region_create()
+
+        # Enable the region and set it to the default navigation map.
+        NavigationServer3D.region_set_enabled(region_rid, true)
+        NavigationServer3D.region_set_map(region_rid, get_world_3d().get_navigation_map())
+
+        # Some mega-nodes like GridMap are often not ready on the first frame.
+        # Also the parsing needs to happen on the main-thread.
+        # So do a deferred call to avoid common parsing issues.
+        call_deferred("parse_source_geometry")
```
Comment (timothyqiu) [inline tutorials/navigation/navigation_using_navigationmeshes.rst:324]:
```suggestion
        parse_source_geometry.call_deferred()
```

Same as the 2D version.

**Review by skyace65 — APPROVED** [review]
Everything looks good to me!

**Review by mhilbrunner — APPROVED** [review]

**Comment by mhilbrunner** [comment]
The `TileMap` thing seems like a code smell on our side, but I guess its better to document it for now instead of omitting it. Looks good to me too! If nothing else comes up, I'll merge this tomorrow. :) Good work, and thanks for keeping the navigation docs up to date with real-life solutions.

**Comment by smix8** [comment]
> The TileMap thing seems like a code smell on our side,

What parts are you referring to?

File: tutorials/navigation/navigation_using_navigationmeshes.rst [hunk 6]
```diff
@@ -229,95 +229,152 @@ There are some common user problems and important caveats to consider when creat
...
+        region_rid = NavigationServer3D.region_create()
+
+        # Enable the region and set it to the default navigation map.
+        NavigationServer3D.region_set_enabled(region_rid, true)
+        NavigationServer3D.region_set_map(region_rid, get_world_3d().get_navigation_map())
+
+        # Some mega-nodes like GridMap are often not ready on the first frame.
+        # Also the parsing needs to happen on the main-thread.
+        # So do a deferred call to avoid common parsing issues.
+        parse_source_geometry.call_deferred()
```
Comment (mhilbrunner) [inline tutorials/navigation/navigation_using_navigationmeshes.rst:324]: This, specifically. Again, better to document this than not, but maybe something we should file somewhere to see if we can improve UX here.

**Comment by mhilbrunner** [comment]
Thanks!

**Comment by mhilbrunner** [comment]
Cherry-picked to 4.2.

**Pull request merged**